irony-cdb.el

```diff
@@ -114,6 +128,24 @@ for files that it cannot handle."
 ;; Functions
 ;;
 
+(defun irony-cdb--locate-dominating-file-with-dirs (file name subdirectories)
+  "Convenience wrapper around `locate-dominating-file'
+
+Looks up the directory hierarchy from FILE for to locate any directory
+in `subdirectories` which contains NAME. Stops at the first parent
+directory where the test succeeds and returns the full path to file
+found."
+  (let ((check-subdirs
```

Comment:
> How do we choose which one to use? First, in most cases there'll be at most one so this isn't critical. If there are more then 1 we can choose the one with the most delimiters as a good heuristic for "most specific". Or we could just take the first one and punt on the whole thing.

Not sure to understand "the most delimiters".

> Which do you prefer?

I think you got the logic right in a previous comment:

> > 1. No. the behavior is to find the matching subdirectory in the nearest parent that matches.
if I (locate-dominating dir ".") and locate ../../<>.json 2 levels up, It won't find the nearer ../build/<>.json. Doing it this way finds the closest one and potentially performs less checks overall.

We have to find the closest one. Sub-dir when composed should be preferred to path closest to the root in the hierarchy.

I mean if:

```el
 (setq irony-cdb-search-directory-list ("." "build" "out/x86_64"))
```
Assuming `default-directory` / `buffer-file-name directory` is `~/foo/bar/qux.cpp`:

1. `~/foo/bar/out/x86_64/compile_commands.json`
2. `~/foo/build/compile_commands.json`
3. `~/compile_commands.json`